airbyte-commons/src/main/java/io/airbyte/commons/util/ResourceIterators.java

```diff
@@ -0,0 +1,186 @@
+import io.airbyte.commons.concurrency.VoidCallableNoException;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public class ResourceIterators {
+
+  public static <T> ResourceIterator<T> emptyIterator() {
+    return new DefaultResourceIterator<>(Collections.emptyIterator(), VoidCallable.NOOP);
+  }
+
+  /**
+   * Coerces a vanilla {@link Iterator} into a {@link ResourceIterator} by adding a no op close
+   * function.
+   *
+   * @param iterator iterator to convert
+   * @param <T> type
+   * @return closeable iterator
+   */
+  public static <T> ResourceIterator<T> resourceIterator(Iterator<T> iterator) {
+    return new DefaultResourceIterator<>(iterator, VoidCallable.NOOP);
+  }
+
+  /**
+   * Coerces a vanilla {@link Iterator} into a {@link ResourceIterator}. The provided {@param onClose}
+   * function will be called at most one time.
+   *
+   * @param iterator resource iterator to add another close to
+   * @param onClose the function that will be called on close
+   * @param <T> type
+   * @return new resource iterator with the close function appended
+   */
+  public static <T> ResourceIterator<T> resourceIterator(Iterator<T> iterator, VoidCallableNoException onClose) {
+    return new DefaultResourceIterator<>(iterator, onClose::call);
+  }
+
+  /**
+   * Wraps a {@link Stream} in a {@link ResourceIterator}. The first time
+   * {@link ResourceIterator#close()} is called, {@link Stream#close()} will be called. It will not be
+   * called again subsequently.
+   *
+   * @param stream stream to wrap
+   * @param <T> type
+   * @return resource iterator
+   */
+  public static <T> ResourceIterator<T> resourceIterator(Stream<T> stream) {
+    return new DefaultResourceIterator<>(stream.iterator(), stream::close);
+  }
+
+  /**
+   * Returns a {@link ResourceIterator} that will call the provided supplier ONE time when
+   * {@link ResourceIterator#hasNext()} is called the first time. The supplier returns a stream that
+   * will be exposed as an iterator.
+   *
+   * @param iteratorSupplier supplier that provides a resource iterator that will be invoked lazily
+   * @param <T> type
+   * @return resource iterator
+   */
+  public static <T> ResourceIterator<T> lazyResourceIterator(Supplier<ResourceIterator<T>> iteratorSupplier) {
+    return new LazyResourceIterator<>(iteratorSupplier);
+  }
+
+  /**
+   * Returns a {@link ResourceIterator} that will call {@link ResourceIterator#close()} at most one
+   * time. Either as soon as {@link Iterator#hasNext()} returns false for the first time or when
+   * {@link ResourceIterator#close()} is called,
+   *
+   * @param resourceIterator wrapped resource iterator
+   * @param <T> type
+   * @return resource iterator
+   */
+  public static <T> ResourceIterator<T> autoClosingResourceIterator(ResourceIterator<T> resourceIterator) {
```

Comment:
This function really hard to understand because the input type and the return type are the same.

Reply:
I think you can make it more obvious by only having one kind of iterator: `AutoCloseableIterator` which offers:
1. Closes when it reaches the end of the stream
2. Closes when you call close

Reply:
You are suggesting an additional interface?

Reply:
To make sure I understand what you're suggesting... I think you are describing the existing `ResourceIterator ` interface. So you would like me to add another interface and called `AutoCloseableIterator` which is identical (either through copy past or inheritance) to `ResourceIterator` ?

Reply:
Or maybe you're just not like the use of composition. We can just discuss on Monday maybe.

Reply:
This is just the decorator pattern, right? Would it help with clarity if the method were called `decorateWithAutoClose` ?

Reply:
gone.